core/arch/arm/mm/core_mmu.c

```diff
@@ -837,12 +843,14 @@ static void init_mem_map(struct tee_mmap_region *memory_map, size_t num_elems)
 	assert(va >= CFG_TEE_RAM_START);
 	assert(end <= CFG_TEE_RAM_START + CFG_TEE_RAM_VA_SIZE);
 
-	add_pager_vaspace(memory_map, num_elems, va, end, &last);
+	add_pager_vaspace(memory_map, num_elems, va, end, &last,
+			  &next_pgdir_va);
 
 	if (core_mmu_place_tee_ram_at_top(va)) {
 		/* Map non-flat mapped addresses below flat mapped addresses */
 		for (map = memory_map; !core_mmap_is_end_of_table(map); map++) {
-			if (map_is_flat_mapped(map))
+			if (map_is_flat_mapped(map) ||
+			    map->type == MEM_AREA_PAGER_VASPACE)
```

Comment:
Maybe we could replace this with `if (map->va) continue;` so that only the areas without an assigned virtual address get one assigned.